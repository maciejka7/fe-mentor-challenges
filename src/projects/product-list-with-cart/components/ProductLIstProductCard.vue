<script setup lang="ts">
import type { Product } from '../types'
import ProductListAddToCartButton from './ProductListAddToCartButton.vue'

const { product } = defineProps<{ product: Product }>()

const breakpoints = useBreakpoints({
  mobile: 0, // optional
  tablet: 640,
  desktop: 1024,
})

const activeBreakpoint = breakpoints.active()

const imageUrl = computed(() => {
  const urlString: string = String(product.image[activeBreakpoint.value!])
  return `/src/projects/product-list-with-cart${urlString.slice(1, urlString.length)}`
})
</script>

<template>
  <div class="max-w-60 rounded-xl">
    <div data-test="card-container" class="relative w-full">
      <img :src="imageUrl" class="rounded-xl" :alt="product.name">
      <ProductListAddToCartButton class="absolute left-1/2 w-fit -top-6 -translate-x-1/2" label="Add to cart" />
    </div>
    <p class="text-plwc-rose-500 text-sm -mt-3" data-test="product-category">
      {{ product.category }}
    </p>
    <h2 class="text-plwc-rose-900 text-lg font-bold" data-test="product-name">
      {{ product.name }}
    </h2>
    <p class="text-plwc-red text-md font-bold" data-test="product-price">
      {{ product.price.money() }}
    </p>
  </div>
</template>

<style scoped></style>
