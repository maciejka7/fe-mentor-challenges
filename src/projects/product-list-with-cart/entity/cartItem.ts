import type { Product } from '../types'

export interface CartItem {
  item: Product
  quantity: number
}

export interface CartItemEntity extends CartItem {
  equals: (a: CartItem, b: CartItem) => boolean
}

export function createCartItem(cartItem: CartItem): CartItemEntity {
  const equals = (a: CartItem, b: CartItem) => {
    const isNameEqual = a.item.name === b.item.name
    const isCategoryEqual = a.item.category === b.item.category
    const isPriceEqual = a.item.price.money() === b.item.price.money()

    if (isNameEqual && isCategoryEqual && isPriceEqual) {
      return true
    }
    return false
  }

  return {
    item: cartItem.item,
    quantity: cartItem.quantity,
    equals,
  }
}
