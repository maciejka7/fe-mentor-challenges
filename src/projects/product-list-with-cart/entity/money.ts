import type { Money } from '../types'

export function createMoney(amount: number, sign = '$'): Money {
  const [base, rest] = String(amount).split('.')

  const createMoneyString = (): string => {
    const restOfValue = rest === undefined ? '00' : rest.length === 1 ? `${rest}0` : `${rest}`
    return `${sign}${base}.${restOfValue}`
  }

  return {
    amount,
    sign,
    money: () => createMoneyString(),
  }
}
