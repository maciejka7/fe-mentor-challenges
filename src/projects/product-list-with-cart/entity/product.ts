import type { ApiProductData, Product } from '../types'
import { createMoney } from './money'

export function createProduct(
  productApiResponse: ApiProductData,
): Product {
  const { category, image, name, price } = productApiResponse

  const product = {
    category,
    name,
    image,
    price: createMoney(price),
  }
  return product
}
