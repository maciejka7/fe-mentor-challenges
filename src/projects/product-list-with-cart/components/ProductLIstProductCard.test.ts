import { type VueWrapper, mount } from '@vue/test-utils'
import type { Mock } from 'vitest'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { createProduct } from '../entity/product'
import { mockProductCardData } from '../helpers/test/testData'
import ProductLIstProductCard from './ProductLIstProductCard.vue'

const cardElements = {
  productImage(wrapper: VueWrapper) { return wrapper.get('[data-test="product-image"]') },
  productCategory(wrapper: VueWrapper) { return wrapper.get('[data-test="product-category"]') },
  productName(wrapper: VueWrapper) { return wrapper.get('[data-test="product-name"]') },
  productPrice(wrapper: VueWrapper) { return wrapper.get('[data-test="product-price"]') },
  addToCartButtonIdleState(wrapper: VueWrapper) { return wrapper.get('[data-test="idle-button"]') },
  addToCartButtonActiveState(wrapper: VueWrapper) { return wrapper.get('[data-test="active-button"]') },
  addToCartButtonActiveAddButton(wrapper: VueWrapper) { return wrapper.get('[data-test="active-button-add"]') },

}

vi.mock('@vueuse/core', () => ({
  useBreakpoints: vi.fn(),
}))

let mockProductCard: VueWrapper
describe('productListProductCard test', async () => {
  beforeEach(() => {
    (useBreakpoints as Mock).mockReturnValue({ active: () => 'mobile' })
    mockProductCard = mount(ProductLIstProductCard, { props: { product: createProduct(mockProductCardData[0]) } })
  })

  it('should render product image', async () => {
    expect(mockProductCard.text()).toContain('Waffle')
    expect(mockProductCard.text()).toContain('$')
    expect(mockProductCard.html()).toMatchSnapshot()
  })

  it('should render product category', () => {
    const category = cardElements.productCategory(mockProductCard)
    expect(category.text()).toContain('Waffle')
  })
  it('should render product name', () => {
    const name = cardElements.productName(mockProductCard)
    expect(name.text()).toContain('Waffle')
  })

  it('should render product price', () => {
    const price = cardElements.productPrice(mockProductCard)
    expect(price.text()).toContain('6.50')
    expect(price.text()).toContain('$')
  })

  it('should render add to cart button', () => {
    const addToCartButton = cardElements.addToCartButtonIdleState(mockProductCard)
    expect(addToCartButton.text()).toContain('Add to cart')
  })

  it('should emit events on amount change', async () => {
    const addToCartButton = cardElements.addToCartButtonIdleState(mockProductCard)
    await addToCartButton.trigger('click')

    const addButton = cardElements.addToCartButtonActiveAddButton(mockProductCard)
    await addButton.trigger('click')
    await addButton.trigger('click')
    await addButton.trigger('click')
    const addToCartActiveButton = cardElements.addToCartButtonActiveState(mockProductCard)
    expect(addToCartActiveButton.text()).toContain('3')
  })

  it.todo('should', () => {

  })
})
