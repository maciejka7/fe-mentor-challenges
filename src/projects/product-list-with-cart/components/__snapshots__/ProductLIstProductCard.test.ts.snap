// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`productListProductCard test > should render product image 1`] = `
"<div class="max-w-60 rounded-xl">
  <div data-test="card-container" class="relative w-full"><img src="/src/projects/product-list-with-cartndefined" class="rounded-xl" alt="Waffle with Berries"><button data-test="idle-button" class="border-plwc-rose-400 hover:border-plwc-red hover:text-plwc-red relative min-w-40 w-fit flex gap-2 border rounded-full bg-white px-7 py-3 transition-all transition-duration-300 absolute left-1/2 w-fit -top-6 -translate-x-1/2"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-shopping-cart-icon text-plwc-red">
        <circle cx="8" cy="21" r="1"></circle>
        <circle cx="19" cy="21" r="1"></circle>
        <path d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"></path>
      </svg> Add to cart</button></div>
  <p class="text-plwc-rose-500 text-sm -mt-3" data-test="product-category">Waffle</p>
  <h2 class="text-plwc-rose-900 text-lg font-bold" data-test="product-name">Waffle with Berries</h2>
  <p class="text-plwc-red text-md font-bold" data-test="product-price">$6.50</p>
</div>"
`;
